includes/class-wcsatt-scheme-prices.php

```diff
@@ -1,388 +1,484 @@
+
+		add_filter( 'woocommerce_available_variation', array( __CLASS__, 'filter_variation_data' ), 0, 3 );
+
+		do_action( 'wcsatt_add_price_filters' );
 	}
 
 	/**
-	 * Remove price filters after modifying child product prices depending on the per-product pricing option state, including any discounts defined at bundled item level.
+	 * Remove price filters.
 	 *
 	 * @return  void
 	 */
 	public static function remove_price_filters() {
 
-		if ( self::$price_overriding_scheme ) {
-
+		if ( WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ) {
+			remove_filter( 'woocommerce_product_get_price', array( __CLASS__, 'filter_get_price' ), 0, 2 );
+			remove_filter( 'woocommerce_product_get_sale_price', array( __CLASS__, 'filter_get_sale_price' ), 0, 2 );
+			remove_filter( 'woocommerce_product_get_regular_price', array( __CLASS__, 'filter_get_regular_price' ), 0, 2 );
+			remove_filter( 'woocommerce_product_variation_get_price', array( __CLASS__, 'filter_get_price' ), 0, 2 );
+			remove_filter( 'woocommerce_product_variation_get_sale_price', array( __CLASS__, 'filter_get_sale_price' ), 0, 2 );
+			remove_filter( 'woocommerce_product_variation_get_regular_price', array( __CLASS__, 'filter_get_regular_price' ), 0, 2 );
+		} else {
 			remove_filter( 'woocommerce_get_price', array( __CLASS__, 'filter_get_price' ), 0, 2 );
 			remove_filter( 'woocommerce_get_sale_price', array( __CLASS__, 'filter_get_sale_price' ), 0, 2 );
 			remove_filter( 'woocommerce_get_regular_price', array( __CLASS__, 'filter_get_regular_price' ), 0, 2 );
-			remove_filter( 'woocommerce_variation_prices', array( __CLASS__, 'filter_get_variation_prices' ), 0, 3 );
+		}
 
-			do_action( 'wcsatt_remove_price_filters' );
+		remove_filter( 'woocommerce_get_variation_prices_hash', array( __CLASS__, 'filter_variation_prices_hash' ), 0, 2 );
+		remove_filter( 'woocommerce_variation_prices', array( __CLASS__, 'filter_get_variation_prices' ), 0, 2 );
 
-			self::$price_overriding_scheme  = false;
-			self::$price_overridden_product = false;
-		}
+		remove_filter( 'woocommerce_available_variation', array( __CLASS__, 'filter_variation_data' ), 0, 3 );
+
+		do_action( 'wcsatt_remove_price_filters' );
 	}
 
 	/**
-	 * Filter get_variation_prices() calls to take price overrides into account.
+	 * Add price html filters.
 	 *
-	 * @param  array                $prices_array
-	 * @param  WC_Product_Variable  $product
-	 * @param  boolean              $display
-	 * @return array
+	 * @return  void
 	 */
-	public static function filter_get_variation_prices( $prices_array, $product, $display ) {
+	public static function add_price_html_filters() {
 
-		$subscription_scheme = self::$price_overriding_scheme;
+		add_filter( 'woocommerce_get_price_html', array( __CLASS__, 'filter_get_price_html' ), 9999, 2 );
 
-		if ( $subscription_scheme && self::has_subscription_price_override( $subscription_scheme ) ) {
+		do_action( 'wcsatt_add_price_html_filters' );
```

Comment:
When the time comes to merge (the core parts of) SATT into WCS, we will need to:

1. Run some migration code on the WCS side. Some new data/meta will be created, e.g. as "legacy" subscription product types are converted to "core" types. The new data will be (most likely) prefixed with `_wcs_` -- e.g. `_wcs_subscription_schemes` vs. `_wcsatt_schemes` in SATT v2. We can be as liberal as we want with choosing new names for **meta and filters**.

2. Run some migration code on the SATT side to prepare SATT meta from merging into WCS, e.g. convert `_wcsatt_schemes` meta to `_wcs_subscription_schemes`. Add a **filter deprecator** class in SATT as the only "remnant" of all code ported into WCS, to avoid the overhead of having all of it in WCS. People who used to run SATT before and need full backwards compatibility with WCS 3.0 will need to run the latest SATT version, which will (a) migrate their data and (b) deprecate old filters.

Part (2) effectively **makes SATT fully responsible for the migration of existing SATT data for compatibility with WCS 3.0**, not the other way around. The interaction of WCS 3.0 with SATT should be limited to checking whether it exists and show an admin notice when the installed version is old.

Summing up:

Because lots of DB data already exists in SATT that will need to be migrated,

and;

assuming that the best approach is to eventually make SATT responsible for migrating its own data,

I'd vote in favor of keeping all filter/meta names in SATT agnostic to the eventual "merge" with WCS, to avoid making _any_ migration-related decisions now, and maintain naming consistency.